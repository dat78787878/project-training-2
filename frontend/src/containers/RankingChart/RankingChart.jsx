import React from 'react';
import { Chart, CategoryScale, LinearScale, BarElement, Title, Tooltip, Legend } from 'chart.js';
import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { getData } from '../../redux/rankingChart/actions';
import Error from '../../components/Error/Error';
import { Bar } from 'react-chartjs-2';
import Loading from '../../components/Loading/Loading';
import ChartDataLabels from 'chartjs-plugin-datalabels';
Chart.register(BarElement, CategoryScale, LinearScale, Tooltip, Legend, Title, ChartDataLabels);

const RankingChart = () => {
  const { rankingData, isLoading, isError } = useSelector((state) => state.rankingChart);
  const { fromDate, toDate } = useSelector((state) => state.dateRange);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getData([fromDate, toDate]));
  }, [fromDate, toDate]);

  const options = {
    indexAxis: 'y',
    scales: {
      x: {
        display: false
      },
      y: {
        ticks: {
          display: true,
          padding: 10,
          mirror: true,
          font: {
            size: 16
          }
        }
      }
    },
    elements: {
      bar: {
        borderWidth: 0
      }
    },
    responsive: true,
    plugins: {
      legend: {
        display: false
      },
      title: {
        display: true,
        text: 'Ranking',
        align: 'start',
        font: {
          size: 18,
          weight: 0
        },
        padding: {
          bottom: 20
        }
      },
      datalabels: {
        // This code is used to display data values
        anchor: 'end',
        align: 'center',
        formatter: Math.round,
        font: {
          weight: 'bold',
          size: 16
        }
      }
    }
  };

  const data = {
    labels: rankingData.map((val) => val.key),
    datasets: [
      {
        barPercentage: 1.0,
        categoryPercentage: 1.0,
        data: rankingData.map((val) => val.value),
        backgroundColor: '#683da85c'
      }
    ]
  };

  const plugins = {
    plugins: [ChartDataLabels]
  };

  return (
    <div className="rankingChart">
      {isLoading && <Loading />}
      {isError && <Error />}
      {!isError && !isLoading && <Bar data={data} options={options} plugins={plugins} />}
    </div>
  );
};

export default RankingChart;
