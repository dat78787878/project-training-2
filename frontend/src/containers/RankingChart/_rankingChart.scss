.rankingChart {
  background-color: white;

  @include pc {
    height: 400px;
  }

  @include tablet {
    height: 383px;
  }

  @include laptop {
    height: 370px;
  }

  height: 370px;

  @include laptop-l {
    height: 370px;

    canvas {
      height: 350px !important;
    }
  }

  @include mobile {
    height: 270px;

    canvas {
      height: 260px !important;
    }
  }

  .pr {
    @include laptop {
      padding-right: 20px;
    }

    @include laptop-l {
      padding-right: 20px;
    }
  }
}
