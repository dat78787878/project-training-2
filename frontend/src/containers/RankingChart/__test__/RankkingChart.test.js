import React from 'react';
import { render, screen } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import RankingChart from '../RankingChart';
window.React = React;

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));

describe('RankingChart Component', () => {
  let mockAppState = {};
  beforeEach(() => {
    useDispatch.mockImplementation(() => () => {});
    useSelector.mockImplementation((callback) => {
      return callback(mockAppState);
    });
  });
  afterEach(() => {
    useSelector.mockClear();
    useDispatch.mockClear();
  });
  it('check loading', () => {
    mockAppState = {
      rankingChart: {
        rankingData: [],
        isLoading: true,
        isError: false
      },
      dateRange: {
        fromDate: '08/08/2021',
        toDate: '09/09/2021'
      }
    };
    const { container } = render(<RankingChart />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('div').firstChild.className).toEqual('loading');
  });
  it('success task', () => {
    mockAppState = {
      rankingChart: {
        rankingData: [
          { label: 'Day 1', value: 9 },
          { label: 'Day 2', value: 6 },
          { label: 'Day 3', value: 6 },
          { label: 'Day 4', value: 6 },
          { label: 'Day 5', value: 6 }
        ],
        isLoading: false,
        isError: false
      },
      dateRange: {
        fromDate: '08/08/2021',
        toDate: '09/09/2021'
      }
    };
    const { container } = render(<RankingChart />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('canvas')).toBeTruthy();
  });

  it('error task', () => {
    mockAppState = {
      rankingChart: {
        rankingData: [],
        isLoading: false,
        isError: true
      },
      dateRange: {
        fromDate: '08/08/2021',
        toDate: '09/09/2021'
      }
    };
    const { container } = render(<RankingChart />);
    expect(screen.getByText('Error')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
