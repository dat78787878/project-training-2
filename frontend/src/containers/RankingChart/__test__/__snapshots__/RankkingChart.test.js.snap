// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RankingChart Component check loading 1`] = `
<div>
  <div
    class="rankingChart padding-title"
  >
    <div
      class="loading"
    >
      <div
        class="spinner-border text-success"
        data-testid="loading"
      />
    </div>
  </div>
</div>
`;

exports[`RankingChart Component error task 1`] = `
<div>
  <div
    class="rankingChart padding-title"
  >
    <div
      class="error"
    >
      Error
      <img
        alt="not-found"
        height="100px"
        src="https://www.pngitem.com/pimgs/m/561-5616833_image-not-found-png-not-found-404-png.png"
        width="100px"
      />
    </div>
  </div>
</div>
`;

exports[`RankingChart Component success task 1`] = `
<div>
  <div
    class="rankingChart padding-title"
  >
    <canvas
      height="150"
      role="img"
      width="300"
    />
  </div>
</div>
`;
